### ISSUE TITLE
Provide a reason for a request being rejected as bad by web endpoint infrastructure

### ISSUE THREAD

**Issue by wilkinsona** [spring-projects/spring-boot#10618]
The web endpoint infrastructure may return a bad request (400) response. It does so when a required parameter is missing, or a parameter could not be mapped to the type required by the operation. Rather than just returning a 400 with no further information, we should include something in the response that helps the user to figure out why the request was deemed to be bad.

See #10568 and #8798.

**Comment by snicoll** [comment]
I had a quick look to this and the right way to fix this issue is to use Spring Boot's error handling. There is [at least another place](https://github.com/spring-projects/spring-boot/pull/10588/files#diff-62076d9efd0b0524b3f9e5dfe2e38e8fR92) where we override the `reason` so we should probably harmonize that.

Using the same trick won't work as WebFlux doesn't have support for this at the moment.

**Comment by snicoll** [comment]
I've [been hacking](3cdc5f2) to try to use error handling when a know exception is thrown. It looks like the general error page mechanism (+ json support for machine client) does not work with Jersey so I don't know if we should hack something for those use cases or live with the limitation.

I can't manage to have proper error handling with MVC. I do have error page handling but for some reason it returns HTML rather than JSON. Perhaps content negotiation in the test needs to be fixed.

For WebFlux, I am using `ResponseStatusException` but it's `toString` representation reuses the message so rather than having a `message` attribute with the provided reason, we get this

```
Response status 400 with reason "Failed to invoke operation because the following required parameters were missing: one"
```

Full body

```json
{
  "timestamp": 1515763768641,
  "path": "/endpoints/query",
  "message": "Response status 400 with reason \"Failed to invoke operation because the following required parameters were missing: one\"",
  "status": 400,
  "error": "Bad Request"
}
```

@bclozel this looks like an issue to me as using this exception is one if not the recommended approach.

**Comment by snicoll** [comment]
The WebFlux issue has been fixed in #11614. I still need to figure out what to do for Jersey.

**Comment by philwebb** [comment]
Blocked until the endpoint refactor is done
